src/plugin/ipc/ssh/util_ssh.cpp

```diff
@@ -151,82 +151,95 @@ void client_loop(int ssh_stdin, int ssh_stdout, int ssh_stderr, int sock)
+      fds.push_back(socketFd);
     }
     if (buffer_ready_for_read(&stderr_buffer)) {
-      FD_SET(ssh_stderr, &readset);
+      socketFd.fd = ssh_stderr;
+      socketFd.events = POLLIN;
+      fds.push_back(socketFd);
     }
 
     if (buffer_ready_for_write(&stdin_buffer)) {
-      FD_SET(ssh_stdin, &writeset);
+      socketFd.fd = ssh_stdin;
+      socketFd.events = POLLOUT;
+      fds.push_back(socketFd);
     }
     if (buffer_ready_for_write(&stdout_buffer)) {
-      FD_SET(STDOUT_FILENO, &writeset);
+      socketFd.fd = STDOUT_FILENO;
+      socketFd.events = POLLOUT;
+      fds.push_back(socketFd);
     }
     if (buffer_ready_for_write(&stderr_buffer)) {
-      FD_SET(STDERR_FILENO, &writeset);
+      socketFd.fd = STDERR_FILENO;
+      socketFd.events = POLLOUT;
+      fds.push_back(socketFd);
     }
 
-    int ret = select(max_fd, &readset, &writeset, &errorset, &tv);
+    int ret = poll((struct pollfd*)&fds[0], fds.size(), 10*1000);
     if (ret == -1 && errno == EINTR) {
       continue;
     }
     if (ret == -1) {
-      perror("select failed");
+      perror("poll failed");
       return;
     }
 
     if (quit_pending)
       break;
 
-    //Read from our STDIN or stdout/err of ssh
-    if (FD_ISSET(STDIN_FILENO, &readset)) {
-      buffer_read(&stdin_buffer, STDIN_FILENO);
-    }
-    if (FD_ISSET(ssh_stdout, &readset)) {
-      buffer_read(&stdout_buffer, ssh_stdout);
-    }
-    if (FD_ISSET(ssh_stderr, &readset)) {
-      buffer_read(&stderr_buffer, ssh_stderr);
-    }
-
-    // Write to our stdout/err or stdin of ssh
-    if (FD_ISSET(ssh_stdin, &writeset)) {
-      buffer_write(&stdin_buffer, ssh_stdin);
-    }
-    if (FD_ISSET(STDOUT_FILENO, &writeset)) {
-      buffer_write(&stdout_buffer, STDOUT_FILENO);
-    }
-    if (FD_ISSET(STDERR_FILENO, &writeset)) {
-      buffer_write(&stderr_buffer, STDERR_FILENO);
-    }
-
-    if (FD_ISSET(remoteSock, &errorset)) {
-      break;
+    std::vector<struct pollfd>::iterator it;
+    for (it = fds.begin(); it != fds.end(); it++) {
+      //Read from our STDIN or stdout/err of ssh
+      if (it->fd == STDIN_FILENO && it->revents & POLLIN) {
+        buffer_read(&stdin_buffer, STDIN_FILENO);
+      }
+      if (it->fd == ssh_stdout && it->revents & POLLIN) {
+        buffer_read(&stdout_buffer, ssh_stdout);
+      }
+      if (it->fd == ssh_stderr && it->revents & POLLIN) {
+        buffer_read(&stderr_buffer, ssh_stderr);
+      }
+
+      // Write to our stdout/err or stdin of ssh
+      if (it->fd == ssh_stdin && it->revents & POLLOUT) {
+        buffer_write(&stdin_buffer, ssh_stdin);
+      }
+      if (it->fd == STDOUT_FILENO && it->revents & POLLOUT) {
+        buffer_write(&stdout_buffer, STDOUT_FILENO);
+      }
+      if (it->fd == STDERR_FILENO && it->revents & POLLOUT) {
+        buffer_write(&stderr_buffer, STDERR_FILENO);
+      }
+
+      if (it->fd == remoteSock && (it->revents & (POLLHUP | POLLERR | POLLNVAL))) {
+        goto out;
```

Comment:
Maybe `end` is a more meaningful name than `out`.  Your intention is to go to the end of the function -- not simply to go "out" of several loops.